### ISSUE TITLE
Missing suites filtering in 6.x

### ISSUE THREAD

**Issue by Ftelf** [lukejpreston/xunit-viewer#67]
In 5.x version I can filter failed/passed suites in the report header. This was extremely useful to filter the 120 passing test suites but leave only 2 failing ones. In the 6.x version this feature seems gone or hidden. I can only expand or collapse the suites, not filter them out. And this test stage is only the small one. I would like to use the xunit-viewer for other stage with thousands of test suites. Thank you in advance for consideration to reintroduce that filter.

**Comment by lukejpreston** [comment]
I would be happy to look into this if you could provide some sample XML.

Suites don't technically have a passing/failing status from what I could find, so I decided to remove that feature because it raised more questions of what a failing suite really is. If your XML has this I would love to implement it.

*NOTE TO FUTURE LUKE:* Perhaps a better feature might be to filter out suites with no tests (default behaviour perhaps), that way you could filter out passing tests and you would be left with only failed tests and their suites?

**Comment by Ftelf** [comment]
I have send one job report with xml to your private email, hope it's ok.
Can't understand why would anybody question that filtering on suite level. We found that perfectly logical a reasonable. This was one of the best features ever.
Yes the status report is also on testsuite level in our nightwatchjs junit xmls.
`<testsuites errors="0"
            failures="1"
            tests="10">

  <testsuite name="BVT.PG-2113"
    errors="0" failures="1" hostname="" id="" package="BVT" skipped="8"
    tests="10" time="113.5s" timestamp="Thu, 13 Feb 2020 07:13:47 GMT">
`
There is no need to force that function to everybody, but it there is any chance to enable that back using the command line parameter it would be fabulous.

**Comment by lukejpreston** [comment]
Thanks for the sample, I will have a look using your example XML you sent.

For context, and more for discussion rather than solving the current issue.

Consider the following example `<testsuite failues='1' errors='1'>`

Has it errored or failed? I think before the parser would say the suite has failed, but then you couldn't filter a suite by errored and get this one, hence the conflict. It raises questions of what is worse, an error or a failure and how do you rank common status of tests?

I also chose to ignore these values when parsing because in many examples I came across they were incorrect, typically saying `0` incorrectly and I can infer backwards from the tests themselves.

I have never seen this in a suite though `<testsuite status='failed'>...` which would be useful

**Comment by Ftelf** [comment]
I'll investigate the native nightwatchjs formatter for the discrepancy between failures and errors. To be honest I have no idea what exactly does that means. We are using also a custom jsunit formatter since the original NWJS omits some logging. I'll try to check those points and report back, but it could take some time...

**Comment by Ftelf** [comment]
So, we are using our own junit reporter which is slightly modified nightwatchjs one.
No big changes, just enclosed systemerr in virtual testcase tag since <system-err> was not handled and I'm not sure if it is even in junit definition.
http://fii.cz/vycekd

I went through the nightwatchjs sources to understand errors and failures counter well.
Seems like errors are counted as unhandled exceptions and explicitely thrown Error exceptions

```
 .catch(function(e) {
      self.testResults.errors++;
      throw e;
    });
```

```
TestSuite.prototype.addErrorToResults = function(err) {
  this.testResults.errors++;
  this.testResults.errmessages = [err.message];

  return this;
};
```

and Timeouts
```
doneFn = Utils.setCallbackTimeout(doneFn, hookName, this.asyncHookTimeout, function(err) {
    this.addErrorToResults(err);
```
But failures are assertion problems
```
Object.keys(testSuiteResult.completed).forEach(function(item) {
        if (testSuiteResult.completed[item].failed > 0) {
          failures++;
        }
      });
```
So errors should be really on 0 as far as something unexpected happened. While test failing only on assertions would count only failures and keeps errors === 0.

**Comment by lukejpreston** [comment]
I have done some updates which I hope to solve this issue.

Version `6.1.0` now filters out suites which have no tests or properties. You can use a toggle to show empty suites. So now you should be able to filter by test status and not by suite status.

Until the router work has finished, the steps you would take to get a similar/same view as v5

1. Open `filter` menu
2. expand `tests`
3. click `visible` for `all` which will hide everything
4. click `hidden` for `failure`

You may want to expand `properties` and click `visible` to hide the properties if any of your suites have them.

Once the router work is done you will be able to have these options "pre-toggled" using the query parameters

Please give this a go and let me know what you think of the new feature

**Comment by Ftelf** [comment]
I really appreciate your effort but unfortunately that update does not mimic the experience present in v5.
I see that when I hide the "passed" tests and all properties they became empty and so I can hide "passed" testsuites and see only failed testsuites and test. But there is following differences between how would I do it in v5 and v6.
I would have to write down all failing testsuites and then re-enable again hidden toggles so I would see in the testsuite where the test fails. If some passed or it failed straight away.
I would also need to see the properties so this is also the case that I need to enable them and then search in the list.
Also the testsuites does not change colours as in v5. they are grey no matter if they are passed or failed.

V6 also substitutes test name ie: "BVT.PG-1241" to "BVT PG 1241"
The only problem with V5 I have that there is problem when handling error messages, only stacktrace is visible in the test. Could you please just backport this fix into V5, and we would stick with V5 and will not bother you with our needs :)

```
<failure message="Visual: segmentationView-view-AZ">    at Section.screenshotIdenticalToBaseline [as verifyScreenshot] (/godata/pipelines/DEV-deploy-price-optimization/vendavo-frontend/vendavo-price-optimization/e2e/page/model/segmentationTree.js:22:23)
    at Object.verifyScreenshot [as verifyScreenshotTree] (/godata/pipelines/DEV-deploy-price-optimization/vendavo-frontend/vendavo-price-optimization/e2e/tests/BVT/PG-1241.js:51:29)
    at Object.verifyScreenshotTree [as Enter a search expression &#34;AZ&#34; and validate] (/godata/pipelines/DEV-deploy-price-optimization/vendavo-frontend/vendavo-price-optimization/e2e/tests/BVT/PG-1241.js:103:8)</failure>

    
    </testcase>
```

V5 http://fii.cz/cxrssc
V6 http://fii.cz/myfwujnd

**Comment by lukejpreston** [comment]
Sorry, not sure what you are asking for but does this sum it up?

1. you want the suites to have a status like in `v5` (and also be filterable)
2. you don't like the titles in `v6`
3. you want to update `v5` with better messages

To address these in turn

1. `v6` approach to suites is different from `v5` in that suites are just containers for tests and I haven't come across any XML examples of suites where they have status. I feel like the `v5` logic was flawed and I do not wish to reimplement it.
2. happy to raise an issue and see what people think, I did this for the `raw` and `pretty` output which is why you get `raw` output by default after several users said they wanted this. I found the title case made test titles easier to read but I can see why this might not be preferable
3. Happy to look at a PR if you want to implement `v6` features in `v5` but as per the README I am not planning on adding features to `v5`

---

I am wondering if we can use this text block to show what it is you want to see

Before any filtering

```
[suite 1]
  [passing test]
  [failing test]

[suite 2]
  [failing test]

[suite 3]
  [passing test]
```

After hiding passing teests you would see

```
[suite 1]
  [failing test]

[suite 2]
  [failing test]
```

But what is it you would like to see?

**Comment by Ftelf** [comment]
1. yep. V5 behavior was just perfect for us. Suites have status according to the specs. If any test in suite fails, the suite fails as well.
2. Just really minor, but see no point why to do that.
3. Just adding also the message attribute will fix all our pains
```
<failure message="Visual: segmentationView-view-AZ_Servi">    at Section.screenshotIdenticalToBaseline [as verifyScreenshot] (/godata/pipelines/DEV-deploy-price-optimization/vendavo-frontend/vendavo-price-optimization/e2e/page/model/segmentationTree.js:22:23)
    at Object.verifyScreenshot [as verifyScreenshotTree] (/godata/pipelines/DEV-deploy-price-optimization/vendavo-frontend/vendavo-price-optimization/e2e/tests/BVT/PG-1241.js:51:29)
    at Object.verifyScreenshotTree [as Enter a search expression &#34;AZ Servi&#34; and validate] (/godata/pipelines/DEV-deploy-price-optimization/vendavo-frontend/vendavo-price-optimization/e2e/tests/BVT/PG-1241.js:110:8)</failure>
```
V6 is processing message attribute in failure tag while V5 does not.

The blocks demonstrates perfectly what we would like to achieve.
After hiding passing suites I would see
```
[suite 1]
  [passing test]
  [failing test]

[suite 2]
  [failing test]
```
In our case when we have also properties which contains URL of the recorded test
```
[suite 1]
  [properties]
  [passing test]
  [failing test]

[suite 2]
  [properties]
  [failing test]
```

The use case is that when developer receives notification about failed testStage in our go.cd server he looks at the report and he did not really care about passing suites and tests. What he want to see quickly is the failed ones so he filters out the passing ones and goes to properties one by one and opens the recorded video on another screen. As the url is loading up, he checks the exact test and why it did failed. Right now because of the missing human-readable message he have to go into textlog and find-out the exact log entry (2MB textfile). But still it is a bit faster to look at xunit-viewer first than looking just into the text file.

**Comment by lukejpreston** [comment]
Ah gotchya! makes more sense thanks for the clarification, let me think about this but I would like to see something in `v6` for this

**Comment by zeburek** [comment]
@Ftelf could you install latest version out of repository, Is not it fixes the issue you're getting in?

**Comment by Ftelf** [comment]
@zeburek I think that it's not yet released. I see 6.1.1 as latest version. Anyway I have changed our xml writer to add <failure message attribute which is normally ignored in V5 into the tag innerText so the 5.x version is perfectly OK for us :)
http://fii.cz/ajydqn

**Comment by lukejpreston** [comment]
FYI I have updated to `6.1.2` if you want to test that, I think @Ftelf has a different use case and I have some ideas on how to provide this functionality which I think will prove useful :D

**Comment by Ftelf** [comment]
I have tried the 6.1.4 version and sorry to say, I'm so confused. At very first sight I have no idea how to use it efficiently. I have hacked our xml reporter to copy the message which was not processed in v5 into the innerText of the element so we will stick with the v5.
Here are some or our samples:
https://drive.google.com/drive/folders/1X9KN9Js6fsgqFOXK1DAm1XMYE3VI8f9I?usp=sharing

Also when there are two tests with the same name one is omited in V6.
ie I have two results for PG-000
CHROME_78.0.3904.70_WINDOWS_PG-000.xml
CHROME_80.0.3987.106_Windows_PG-000.xml
V5 shows 2 suites but V6 only one of them.

In V5 everything is so beautifully straighforward:
![image](https://user-images.githubusercontent.com/12624859/75182122-ff89d280-573f-11ea-98fd-6c9ddfa81d28.png)

But in V6
I have no idea what the heck is going on
![image](https://user-images.githubusercontent.com/12624859/75182166-12040c00-5740-11ea-9dee-7fddadd57725.png)

**Comment by lukejpreston** [comment]
I am not sure what you are confused by sorry, from what I can see you have the following duplications in your XML outputs

```
BVT.PG-000
  Login: &#34;tom&#34;
  Landing page -&gt; Draft area -&gt; Open big model
  Bench 0
BVT.PG-000
  Login: &#34;tom&#34;
  Landing page -&gt; Draft area -&gt; Open big model
  Bench 0
```

The only difference between the two suites I can see is one has a failing test `systemerr` but otherwise they are the same. In `v6` the parsing logic concatenates suites and tests with the same names together.

What is it you wanted to see and is this a separate issue than the one you originally raised?

**Comment by lukejpreston** [comment]
I am going to close this issue. From what I can see you have duplicate tests and suites. It looks like the only difference is the filename and some fo the metadata such as time and assertions. If you see any other differences please feel free to re-open and I will be happy to work on it :D
